File: 16.Longest_Repeating_Character Replacement/sol.py
import os
import sys
from typing import List

# ======================================================================
# 🧠 CHALLENGE: Longest Repeating Character Replacement (Python Version)
# ======================================================================
# Description:
# You are given a string s and an integer k. You can choose any character
# of the string and change it to any other uppercase English character.
# You can perform this operation at most k times.
#
# Return the length of the longest substring containing the same letter
# you can get after performing the above operations.
#
# 📋 Rules:
# 1. You may replace at most k characters in a substring.
# 2. After replacements, the substring should be all the same letter.
# 3. Return the maximum possible length.
#
# 💡 Examples:
# - Input: s="ABAB", k=2     => Output: 4
# - Input: s="AABABBA", k=1  => Output: 4
#
# Constraints:
# - 1 <= len(s) <= 1e5
# - s consists of only uppercase English letters
# - 0 <= k <= len(s)
# ======================================================================


# region [📚 Reference Solutions] (Solutions hidden as requested)
def character_replacement1(s: str, k: int) -> int:
    """
    Method: Sliding Window + Frequency Count
    Complexity: Time O(N) | Space O(1)   (since alphabet size is 26)
    """
    freq = [0] * 26
    l = 0
    max_freq = 0  # max frequency of a single char within the current window
    best = 0

    for r, ch in enumerate(s):
        idx = ord(ch) - ord("A")
        freq[idx] += 1
        if freq[idx] > max_freq:
            max_freq = freq[idx]

        # window_size - max_freq = number of replacements needed to unify window
        while (r - l + 1) - max_freq > k:
            left_idx = ord(s[l]) - ord("A")
            freq[left_idx] -= 1
            l += 1

        # now window is valid
        win_len = r - l + 1
        if win_len > best:
            best = win_len

    return best


# endregion


# ======================================================================
#  region [✍️ Practice Area]
#  Please write your solution between the markers below.
# ======================================================================
# <PRACTICE_START>
def character_replacement(s: str, k: int) -> int:
    # TODO: Implement your solution here.
    return 0
# <PRACTICE_END>
# endregion

# ======================================================================
#  region [🚀 Test Runner & Auto-Reset] (Do not modify below this line)
# ======================================================================


def reset_practice_area():
    print("\n🔄 Resetting Practice Area to default state...")
    MARKER_START = "# <PRACTICE_" + "START>"
    MARKER_END = "# <PRACTICE_" + "END>"

    default_code_lines = [
        "def character_replacement(s: str, k: int) -> int:\n",
        "    # TODO: Implement your solution here.\n",
        "    return 0\n",
    ]

    try:
        current_file = __file__
        with open(current_file, "r", encoding="utf-8") as f:
            lines = f.readlines()

        start_idx = -1
        end_idx = -1

        for i, line in enumerate(lines):
            if MARKER_START in line:
                start_idx = i
            if MARKER_END in line:
                end_idx = i

        if start_idx == -1 or end_idx == -1 or start_idx >= end_idx:
            print("⚠️ Error: Markers not found. Reset cancelled.")
            return

        new_content = lines[: start_idx + 1] + default_code_lines + lines[end_idx:]

        with open(current_file, "w", encoding="utf-8") as f:
            f.writelines(new_content)

        print("✨ Reset complete! The file is ready for a fresh start.")
    except Exception as e:
        print(f"⚠️ Reset failed: {e}")


if __name__ == "__main__":
    test_cases = [
        ("ABAB", 2, 4),
        ("AABABBA", 1, 4),
        ("AAAA", 0, 4),
        ("ABCDE", 1, 2),
        ("BAAAB", 2, 5),
        ("", 3, 0),  # not in constraints, but helps robustness
    ]

    print(f"\n🧪 Testing your [character_replacement] function...\n")

    header = f"{'Input (s,k)':<30} | {'Expected':<10} | {'Actual':<10} | Status"
    print(header)
    print("-" * len(header))

    all_pass = True
    for s, k, expected in test_cases:
        try:
            actual = character_replacement(s, k)
        except Exception:
            actual = "Error"

        is_match = actual == expected
        status_icon = "✅ PASS" if is_match else "❌ FAIL"
        if not is_match:
            all_pass = False

        in_str = f'("{s}", {k})'
        if len(in_str) > 28:
            in_str = in_str[:25] + "..."
        print(f"{in_str:<30} | {str(expected):<10} | {str(actual):<10} | {status_icon}")

    print("-" * len(header))
    if all_pass:
        print("\n🎉 Fantastic! All test cases passed.")
        reset_practice_area()
    else:
        print("\n⚠️  Some tests failed. Keep trying! (The file will not reset yet)")
# endregion

File: 16.Longest_Repeating_Character Replacement/sol.go
package main

import (
	"fmt"
	"os"
	"runtime"
	"strings"
)

// ======================================================================
// 🧠 CHALLENGE: Longest Repeating Character Replacement (Go Version)
// ======================================================================
// Description:
// You are given a string s and an integer k. You can choose any character
// of the string and change it to any other uppercase English character.
// You can perform this operation at most k times.
//
// Return the length of the longest substring containing the same letter
// you can get after performing the above operations.
//
// Constraints:
// - 1 <= len(s) <= 1e5
// - s consists of only uppercase English letters
// - 0 <= k <= len(s)
// ======================================================================

// #region [📚 Reference Solutions] (Solutions hidden as requested)
func characterReplacement1(s string, k int) int {
	// Method: Sliding Window + Frequency Count
	// Complexity: Time O(N) | Space O(1) (alphabet size fixed at 26)

	freq := make([]int, 26)
	l := 0
	maxFreq := 0
	best := 0

	for r := 0; r < len(s); r++ {
		idx := int(s[r] - 'A')
		freq[idx]++
		if freq[idx] > maxFreq {
			maxFreq = freq[idx]
		}

		for (r-l+1)-maxFreq > k {
			leftIdx := int(s[l] - 'A')
			freq[leftIdx]--
			l++
		}

		winLen := r - l + 1
		if winLen > best {
			best = winLen
		}
	}

	return best
}

// #endregion

// ======================================================================
//
//	#region [✍️ Practice Area]
//	Please write your solution between the markers below.
//
// ======================================================================
// <PRACTICE_START>
func CharacterReplacement(s string, k int) int {
	// TODO: Implement your solution here.
	return 0
}
// <PRACTICE_END>
// #endregion

// ======================================================================
//  #region [🚀 Test Runner & Auto-Reset] (Do not modify below this line)
// ======================================================================

type TestCase struct {
	s        string
	k        int
	expected int
}

func main() {
	runTests()
}

func runTests() {
	testCases := []TestCase{
		{"ABAB", 2, 4},
		{"AABABBA", 1, 4},
		{"AAAA", 0, 4},
		{"ABCDE", 1, 2},
		{"BAAAB", 2, 5},
		{"", 3, 0}, // not in constraints, but helps robustness
	}

	fmt.Printf("\n🧪 Testing your [CharacterReplacement] function...\n\n")

	header := fmt.Sprintf("%-30s | %-10s | %-10s | %s", "Input (s,k)", "Expected", "Actual", "Status")
	fmt.Println(header)
	fmt.Println(strings.Repeat("-", len(header)))

	allPass := true

	for _, tc := range testCases {
		actual := -999999
		statusIcon := "✅ PASS"

		func() {
			defer func() {
				if r := recover(); r != nil {
					actual = -999999
					statusIcon = "❌ FAIL"
					allPass = false
				}
			}()
			actual = CharacterReplacement(tc.s, tc.k)
		}()

		if actual != tc.expected {
			statusIcon = "❌ FAIL"
			allPass = false
		}

		inStr := fmt.Sprintf("(\"%s\", %d)", tc.s, tc.k)
		if len(inStr) > 28 {
			inStr = inStr[:25] + "..."
		}
		fmt.Printf("%-30s | %-10d | %-10d | %s\n", inStr, tc.expected, actual, statusIcon)
	}

	fmt.Println(strings.Repeat("-", len(header)))

	if allPass {
		fmt.Println("\n🎉 Fantastic! All test cases passed.")
		resetPracticeArea()
	} else {
		fmt.Println("\n⚠️  Some tests failed. Keep trying! (The file will not reset yet)")
	}
}

func resetPracticeArea() {
	fmt.Println("\n🔄 Resetting Practice Area to default state...")

	markerStart := "// <PRACTICE_" + "START>"
	markerEnd := "// <PRACTICE_" + "END>"

	defaultCode := []string{
		"func CharacterReplacement(s string, k int) int {",
		"\t// TODO: Implement your solution here.",
		"\treturn 0",
		"}",
	}

	_, currentFile, _, ok := runtime.Caller(0)
	if !ok {
		fmt.Println("⚠️ Error: Could not determine file path. Reset cancelled.")
		return
	}

	content, err := os.ReadFile(currentFile)
	if err != nil {
		fmt.Printf("⚠️ Error reading file: %v\n", err)
		return
	}

	lines := strings.Split(string(content), "\n")
	startIdx := -1
	endIdx := -1

	for i, line := range lines {
		if strings.Contains(line, markerStart) {
			startIdx = i
		} else if strings.Contains(line, markerEnd) {
			endIdx = i
		}
	}

	if startIdx == -1 || endIdx == -1 || startIdx >= endIdx {
		fmt.Println("⚠️ Error: Markers not found or invalid. Reset cancelled.")
		return
	}

	newLines := make([]string, 0)
	newLines = append(newLines, lines[:startIdx+1]...)
	newLines = append(newLines, defaultCode...)
	newLines = append(newLines, lines[endIdx:]...)

	output := strings.Join(newLines, "\n")
	err = os.WriteFile(currentFile, []byte(output), 0644)
	if err != nil {
		fmt.Printf("⚠️ Error writing file: %v\n", err)
		return
	}

	fmt.Println("✨ Reset complete! The file is ready for a fresh start.")
}

// #endregion
